[FEATURE] Show how to override FSC templates (#287)
TYPO3-Documentation/TYPO3CMS-Tutorial-SitePackage: Documentation/ContentElementRendering/Index.rst

:navigation-title: Content Element rendering
..  include:: /Includes.rst.txt

..  _content-element-rendering:

====================================================
Overriding the default templates of content elements
====================================================

The content elements that are rendered up to this point are rendered by the
TYPO3 system extension Fluid Styled Content (:composer:`typo3/cms-fluid-styled-content`).

This extensions offers default templates to render content elements. Without
such an extension, no content would be rendered at all. The default
templates provided by this extension can be overridden with site settings
provided by Fluid Styled Content.

..  contents::

..  _content-element-rendering-settings:

Use Settings to override template paths of Fluid Styled Content
===============================================================

Site settings can be saved both in the site configuration and in the site
package extension.

We will save the settings to the site package but use the settings editor to

> write the YAML for us

Reviewer: ```suggestion
write the YAML for us.
```

templates:

..  _content-element-rendering-menu-subpages:

Override the "Menu of Subpages" template
========================================

On "Page 1" of the example data a content element of type "Subpages" was added.
We now have to find out what this type is called in the database. The raw values
saved to the database are displayed in the TYPO3 backend when the debug mode
is activated:

..  figure:: /Images/MenuContentElement.png
    :alt: Screenshot demonstrating the backend debug mode for Content element "Subpages" on "Page 1"

    You can now see that the Type (saved in field `CType`) is stored as `menu_subpages`

Now we must find and copy the original template from Fluid Styled Content. TYPO3
extensions are saved by their composer key, here :composer:`typo3/cms-fluid-styled-content`,
into the folder :path:`vendor` during installation via Composer. You can find
the files belonging to Fluid Styled Content in folder
:path:`vendor/typo3/cms-fluid-styled-content` therefore. This folder is
structured similarly to your site package extension and you can find the original
templates in folder :path:`Resources/Private/Templates` here.

> The templates have the name of the `CType` in CamelCase. Copy file

Reviewer: This is a very very simplified interpretation and does not respect the true mapping. Sure, it may work for TYPO3 fluid_styled_content, but it may change every time and you can not expect people to have same mapping understanding in other extensions.

On the one hand your explanation is very easy to understand and helpful, but on the other hand it's nothing to relate on for other extensions. Hmmm....

Example on how the real mapping happens:

```
tt_content.menu_recently_updated =< lib.contentElement
tt_content.menu_recently_updated {
    templateName = MenuRecentlyUpdated
```

Author: Ok, I added a restriction by conventions and templates of Fluid styled content.

> The templates have the name of the `CType` in CamelCase. Copy file

Author: ```suggestion
By convention the templates of Fluid Styled Content have the name of the `CType` in CamelCase. Copy file
```

:navigation-title: Content Element rendering
..  include:: /Includes.rst.txt

..  _content-element-rendering:

====================================================
Overriding the default templates of content elements
====================================================

The content elements that are rendered up to this point are rendered by the
TYPO3 system extension Fluid Styled Content (:composer:`typo3/cms-fluid-styled-content`).

This extensions offers default templates to render content elements. Without
this such an extension, no content would be rendered at all. The default
templates provided by this extension can be overridden with site settings
provided by the Fluid Styled Content.

..  contents::

..  _content-element-rendering-settings:

Use Settings to override template paths of Fluid Styled Content
===============================================================

Site settings can be saved both in the site configuration and in the site
package extension.

> If you would share your site package across several sites you could set defaults

Reviewer 2: I would leave this detail out for a general sitepackage tutorial.

Author: Done

in your site package and override them in your site configuration.

We will save the settings to the site package but use the settings editor to
write the YAML for us

> Got to module :guilabel:`Site Management > Settings` and edit the settings of

Reviewer 2: ```suggestion
Go to module :guilabel:`Site Management > Settings` and edit the settings of
```

your site. Override the paths to the templates of Fluid Styled Content like this:

..  figure:: /Images/SiteSettingsFsc.png
    :alt: Screenshot demonstrating the site settings module

    Override the templates of Fluid Styled Content

If you would click :guilabel:`Save` now, the settings would be saved to your
site configuration at :path:`config/sites/my-site/settings.yaml`. We however

> want to save the settings to the set of our site package extension.

Reviewer 2: ```suggestion
want to save the settings to the site set of our site package extension.
```

Click the button :guilabel:`YAML export` to copy the configuration to your
Clipboard instead, then save it to the following file:

..  code-block:: yaml

>     :caption: packages/site_package/my_site_package/Configuration/Sets/SitePackage/settings.yaml

Reviewer 2: ```suggestion
    :caption: packages/site_package/Configuration/Sets/SitePackage/settings.yaml
```

        layoutRootPath: 'EXT:site_package/Resources/Private/ContentElements/Layouts'

Then close the site settings without saving. You can now start overriding the
templates:

..  _content-element-rendering-menu-subpages:

Override the "Menu of Subpages" template
========================================

On "Page 1" of the example data a content element of type "Subpages" was added.
We now have to find out what this type is called in the database. The raw values
saved to the database are displayed in the TYPO3 backend when the debug mode
is activated:

..  figure:: /Images/MenuContentElement.png
    :alt: Screenshot demonstrating the backend debug mode for Content element "Subpages" on "Page 1"

    You can now see that the Type (saved in field `CType`) is stored as `menu_subpages`

Now we must find and copy the original template from Fluid Styled Content. TYPO3
extensions are saved by their composer key, here :composer:`typo3/cms-fluid-styled-content`,
into the folder :path:`vendor` during installation via Composer. You can find
the files belonging to Fluid Styled Content in folder
:path:`vendor/typo3/cms-fluid-styled-content` therefore. This folder is

> structure similar to your site package extension and you can find the original

Reviewer 2: ```suggestion
structured similarly to your site package extension and you can find the original
```

templates in folder :path:`Resources/Private/Templates` here.

The templates have the name of the `CType` in CamelCase. Copy file
:path:`vendor/typo3/cms-fluid-styled-content/Resources/Private/Templates/MenuSubpages.html`
into folder :path:`packages/my_site_package/Resources/Private/ContentElements/Templates/MenuSubpages.html`

Edit the file to add some classes as used in menus in Bootstrap, for example
like this:

..  include:: /CodeSnippets/Fluid/MenuSubpages.rst.txt

In most parts the changes we made are pretty straight forward. In line 9
we use the `Fluid inline notation <https://docs.typo3.org/permalink/t3coreapi:fluid-inline-notation>`_
of the :ref:`If ViewHelper <f:if> <t3viewhelper:typo3fluid-fluid-if>` to only
output class `active` if the page in the menu is in the root line.

..  _content-element-rendering-sitemap:

Override the sitemap template
=============================

In a similar fashion we now copy and adjust the template for the sitemap from
:path:`vendor/typo3/cms-fluid-styled-content/Resources/Private/Templates/MenuSitemap.html`

> into folder :path:`packages/my_site_package/Resources/Private/ContentElements/Templates/MenuSitemap.html`

Reviewer 2: ```suggestion
into folder :path:`packages/site_package/Resources/Private/ContentElements/Templates/MenuSitemap.html`
```
